File: windows-app/src/ThoughtInput.Core/SettingsBackup.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ThoughtInput.Core;

public sealed record SettingsBackup(
    [property: JsonPropertyName("schemaVersion")] int SchemaVersion,
    [property: JsonPropertyName("exportedAt")] DateTimeOffset ExportedAt,
    [property: JsonPropertyName("sourcePlatform")] string SourcePlatform,
    [property: JsonPropertyName("destinations")] IReadOnlyList<BackupDestination> Destinations)
{
    public const int CurrentSchemaVersion = 1;
}

public sealed record BackupDestination(
    [property: JsonPropertyName("destination")] Destination Destination,
    [property: JsonPropertyName("secrets")] IReadOnlyDictionary<string, string> Secrets);

public sealed class SettingsBackupException(string message) : Exception(message);

public static class SettingsBackupService
{
    public static byte[] Export(DestinationStore store, ISecretStore secrets)
    {
        var backups = store.Destinations.Select(d => new BackupDestination(
            Destination: d,
            Secrets: d.KeychainRefs
                .Select(r => (r.Account, Value: secrets.LoadString(r.Account)))
                .Where(t => t.Value is not null)
                .ToDictionary(t => t.Account, t => t.Value!))).ToList();

        var backup = new SettingsBackup(
            SchemaVersion: SettingsBackup.CurrentSchemaVersion,
            ExportedAt: DateTimeOffset.UtcNow,
            SourcePlatform: CapturePayload.SourcePlatformWindows,
            Destinations: backups);

        return JsonSerializer.SerializeToUtf8Bytes(backup, DestinationJson.BackupOptions);
    }

    public static void Import(byte[] data, DestinationStore store, ISecretStore secrets)
    {
        SettingsBackup? backup;
        try
        {
            backup = JsonSerializer.Deserialize<SettingsBackup>(data, DestinationJson.DefaultOptions);
        }
        catch (Exception ex)
        {
            throw new SettingsBackupException($"Couldn't read the backup file: {ex.Message}");
        }
        if (backup is null)
            throw new SettingsBackupException("Backup file was empty");

        if (backup.SchemaVersion != SettingsBackup.CurrentSchemaVersion)
            throw new SettingsBackupException(
                $"Unsupported backup schema version ({backup.SchemaVersion}). This file was made by a newer version of the app.");

        Guid? activeId = null;
        var newDestinations = new List<Destination>();
        foreach (var entry in backup.Destinations)
        {
            foreach (var (account, value) in entry.Secrets)
                secrets.SaveString(account, value);

            var clone = new Destination
            {
                Id = entry.Destination.Id,
                Name = entry.Destination.Name,
                Type = entry.Destination.Type,
                IsActive = false,
            };
            if (entry.Destination.IsActive) activeId = entry.Destination.Id;
            newDestinations.Add(clone);
        }

        store.ReplaceAll(newDestinations);

        if (activeId is { } id)
        {
            var active = store.Destinations.FirstOrDefault(d => d.Id == id);
            if (active is not null) store.SetActive(active);
        }
    }
}

File: android-app/app/src/main/java/com/thoughtinput/capture/data/SettingsBackup.kt
package com.thoughtinput.capture.data

import com.thoughtinput.capture.data.destinations.Destination
import com.thoughtinput.capture.data.destinations.DestinationStore
import com.thoughtinput.capture.data.destinations.KeychainRef
import com.thoughtinput.capture.data.destinations.SecretsKeystore
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.time.Instant

@Serializable
data class SettingsBackup(
    val schemaVersion: Int = CURRENT_SCHEMA_VERSION,
    val exportedAt: String,
    val sourcePlatform: String = "android",
    val destinations: List<BackupDestination>
) {
    companion object {
        const val CURRENT_SCHEMA_VERSION = 1
    }
}

@Serializable
data class BackupDestination(
    val destination: Destination,
    val secrets: Map<String, String>
)

class SettingsBackupException(message: String, cause: Throwable? = null) : Exception(message, cause)

class SettingsBackupService(
    private val store: DestinationStore,
    private val keystore: SecretsKeystore
) {
    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
        prettyPrint = true
    }

    fun export(): String {
        val backups = store.destinations.value.map { destination ->
            val secrets = destination.type.keychainRefs
                .mapNotNull { ref -> keystore.loadString(ref)?.let { ref.account to it } }
                .toMap()
            BackupDestination(destination = destination, secrets = secrets)
        }
        val backup = SettingsBackup(
            exportedAt = Instant.now().toString(),
            destinations = backups
        )
        return json.encodeToString(SettingsBackup.serializer(), backup)
    }

    fun import(text: String) {
        val backup = try {
            json.decodeFromString(SettingsBackup.serializer(), text)
        } catch (e: Exception) {
            throw SettingsBackupException("Couldn't read the backup file: ${e.message}", e)
        }

        if (backup.schemaVersion != SettingsBackup.CURRENT_SCHEMA_VERSION) {
            throw SettingsBackupException(
                "Unsupported backup schema version (${backup.schemaVersion}). This file was made by a newer version of the app."
            )
        }

        for (existing in store.destinations.value.toList()) {
            store.delete(existing)
        }

        var activeId: java.util.UUID? = null
        for (entry in backup.destinations) {
            for ((account, value) in entry.secrets) {
                keystore.save(KeychainRef(account), value)
            }
            val wasActive = entry.destination.isActive
            store.add(entry.destination.copy(isActive = false))
            if (wasActive) activeId = entry.destination.id
        }

        if (activeId != null) {
            store.destinations.value.firstOrNull { it.id == activeId }?.let { store.setActive(it) }
        }
    }
}
